File: easyfamily-ios/EasyfamilyApp/Features/Finance/AssetViewModel.swift
import Foundation

@MainActor
final class AssetViewModel: ObservableObject {
    @Published var assets: [AssetItem] = []
    @Published var totalValue: Decimal = 0
    @Published var isLoading: Bool = false
    @Published var error: String?

    func load(token: String) async {
        isLoading = true
        error = nil
        do {
            let response = try await APIService.listAssets(token: token)
            assets = response.items
            totalValue = response.totalValue
        } catch {
            self.error = "加载失败：\(error.localizedDescription)"
        }
        isLoading = false
    }

    func create(token: String, name: String, assetType: String, value: Decimal, note: String?) async throws {
        let req = AssetCreateRequest(name: name, assetType: assetType, value: value, note: note?.isEmpty == false ? note : nil)
        _ = try await APIService.createAsset(token: token, req: req)
        await load(token: token)
    }

    func update(token: String, id: Int, name: String, assetType: String, value: Decimal, note: String?) async throws {
        let req = AssetCreateRequest(name: name, assetType: assetType, value: value, note: note?.isEmpty == false ? note : nil)
        _ = try await APIService.updateAsset(token: token, id: id, req: req)
        await load(token: token)
    }

    func delete(token: String, id: Int) async throws {
        try await APIService.deleteAsset(token: token, id: id)
        await load(token: token)
    }
}

File: easyfamily-android/app/src/main/java/com/easyfamily/ui/asset/AssetViewModel.kt
package com.easyfamily.ui.asset

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.easyfamily.data.ApiClient
import com.easyfamily.data.AssetData
import com.easyfamily.data.local.AuthDataStore
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject

data class AssetUiState(
    val assets: List<AssetData> = emptyList(),
    val totalValue: Double = 0.0,
    val loading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class AssetViewModel @Inject constructor(
    private val authDataStore: AuthDataStore
) : ViewModel() {

    private val _uiState = MutableStateFlow(AssetUiState())
    val uiState: StateFlow<AssetUiState> = _uiState.asStateFlow()

    private suspend fun token(): String = authDataStore.accessToken.first()

    fun load() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(loading = true, error = null)
            try {
                val t = token()
                val (items, total) = ApiClient.listAssets(t)
                _uiState.value = _uiState.value.copy(assets = items, totalValue = total, loading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(loading = false, error = e.message)
            }
        }
    }

    fun createAsset(name: String, assetType: String, value: Double, note: String?) {
        viewModelScope.launch {
            try {
                ApiClient.createAsset(token(), name, assetType, value, note)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    fun updateAsset(id: Int, name: String, assetType: String, value: Double, note: String?) {
        viewModelScope.launch {
            try {
                ApiClient.updateAsset(token(), id, name, assetType, value, note)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    fun deleteAsset(id: Int) {
        viewModelScope.launch {
            try {
                ApiClient.deleteAsset(token(), id)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }
}
